Adds Batching Proposal (#307)
APIs-guru/graphql-over-http: spec/Appendix B -- Batching.md

## B. Appendix: Batching

This appendix defines optional batching extensions for GraphQL over HTTP. It
covers both:

- **Variable batching**: one request object executes one operation multiple
  times by using a list of variable maps.
- **Request batching**: one HTTP request contains a list of request objects.

Variable batching and request batching MAY be used independently or together.

Note: GraphQL subscriptions are beyond the scope of this specification, so this
appendix defines batching semantics for query and mutation operations only.

### Motivation

Field-level batching can solve some cases, but it does not cover all shapes of

> variable-dependent queries. For example:

Reviewer: "field-level batching" is not well defined.

```graphql example
query ($vendorId: ID!, $productId: ID!) {
  vendor(id: $vendorId) {
    id
    name
    stock(productId: $productId) {
      count
      nextDeliveryDue
      product {
        id
        name
      }
    }
  }
}
```

A list of variable maps allows this operation to execute multiple times in one

> HTTP request, each time with a different pair of variable values.

Reviewer: I didn't find this wording obvious, but I'm not a big fan of my suggested replacement either:

```suggestion
Supplying the document to execute alongside a _list_ of {variableValues} maps
allows this operation to execute multiple times in a single HTTP request, each
time with a different {variableValues} map.
```

### Request Shapes

A batching request is sent as a POST request with a JSON-encoded body.

This appendix extends the `{variables}` request parameter:

- `{variables}` MAY be a map (as defined in the main specification).
- `{variables}` MAY be a list of maps for variable batching.

> - If `{variables}` is a list, each item in the list MUST be a map.

Reviewer: ```suggestion
To add support for variable batching, the {variables} _request parameter_ is extended such that it may now also accept a list of maps.
```

A batching request body MUST be one of the following:

- A single _GraphQL-over-HTTP request_ object.

> - A list of _GraphQL-over-HTTP request_ objects.

Reviewer: ```suggestion
To add support for request batching, the JSON encoding of a _GraphQL-over-HTTP
request_ is extended such that the body of the request may also be a list of _GraphQL-over-HTTP request_ maps.
```

> If a request object's `{variables}` value is a list, the server MUST execute
> that request once per variables entry.
> 
> If the request body is a list, each entry in that list MUST be processed as an
> independent _GraphQL-over-HTTP request_.

Reviewer: We need to encourage them to discard requests where the total number of requests to execute is over some threshold. But also, I think this is more part of execution behavior than request shape.

```suggestion
```

Author: If the server is allowed to discard requests then this must be specified in some way. Like it cannot simply discard requests in the middle without any notice. Otherwise we need another marker to pin which requests have to be reissued. I a server like executes 20 requests and then says limit reached a stop now that could be handled... although to be honest I would find it better if the server would simply reject it all and say i can only process 20 requests at a time so i reject your batch.

This ties btw back in with capabilities. In the gateway case the gateway could inspect the capabilities and send exactly the correct amount of requests in a batch.

Reviewer: Yes, I think all or none - it should decide up front if it's attempting them or not.

That said, mutations would be serialized and may hit timeouts, and the side effects might not be able to be undone, so... :grimacing:  I guess just execute them all and the later ones produce errors...

... We really need to add some standard error codes to GraphQL errors.

### Execution Semantics

When a request object's `{variables}` value is a list, each variables map

> represents a distinct execution of that request object.
> 
> The server MAY execute these executions in any order, including concurrently,
> and the client MUST NOT assume serial execution or any ordering based on
> variables list position.

Reviewer: ```suggestion
When a request object's `{variables}` value is a list, each variables map
represents a distinct execution of that request object.
The server MAY execute these executions in any order, including concurrently,
and the client MUST NOT assume serial execution or any ordering based on
variables list position.
```
